### ISSUE TITLE
ANSI color in shell prompt seems to break line-wrap

### ISSUE THREAD

**Issue by artyeffem** [kovidgoyal/kitty#8110]
**Describe the bug**
An shell prompt that includes ANSI color codes will prevent long command lines from wrapping to next line

**To Reproduce**
1. export PS1="\033[92m(green-test)\033[0m ${PS1:-}"
2. type until past right-edge of terminal
3. Observe keys overwriting current shell prompt line

**Environment details**
kitty 0.21.2 created by Kovid Goyal
Linux balthasar 5.15.0-113-generic #123-Ubuntu SMP Mon Jun 10 08:16:17 UTC 2024 x86_64
Ubuntu 22.04.5 LTS \n \l
DISTRIB_ID=Ubuntu
DISTRIB_RELEASE=22.04
DISTRIB_CODENAME=jammy
DISTRIB_DESCRIPTION="Ubuntu 22.04.5 LTS"
Running under:X11
Loaded config files:
  /home/aeffem/.config/kitty/kitty.conf

Config options different from defaults:
tab_bar_edge          1
tab_bar_style         separator
update_check_interval 0.0
Added shortcuts:
	ctrl+f4 KeyAction('combine', (KeyAction('new_window'), KeyAction('detach_window')))
	shift+ctrl+i KeyAction('new_os_window_with_cwd')
Changed shortcuts:
	shift+ctrl+e KeyAction('combine', (KeyAction('new_window'), KeyAction('goto_layout', ('tall',))))
	shift+ctrl+o KeyAction('combine', (KeyAction('new_window'), KeyAction('goto_layout', ('fat',))))

**Additional context**
Try to reproduce the problem with `kitty --config NONE` 
_Same results_

**Comment by kovidgoyal** [comment]
Does not reproduce for me with up to date kitty. You are reporting a bug aginst multiple year old versions, update.

**Comment by ssimono** [comment]
Hello,

I am facing the same issue with `0.42.1` which is the most recent version. I confirm this happens only when the prompt has some ANSI colors. I am attaching a screencast of me reproducing it.

https://github.com/user-attachments/assets/af565cca-550d-485a-a35b-be5936e05662

<details><summary>My debug output</summary>
<p>
<pre>
kitty 0.42.1 created by Kovid Goyal
Linux [...] 6.12.31 #1-NixOS SMP PREEMPT_DYNAMIC Thu May 29 09:03:27 UTC 2025 x86_64

<<< Welcome to NixOS 25.05.20250604.8f1b52b (x86_64) - /dev/tty >>>

Run 'nixos-help' for the NixOS manual.

DISTRIB_CODENAME=warbler
DISTRIB_DESCRIPTION="NixOS 25.05 (Warbler)"
DISTRIB_ID=nixos
DISTRIB_RELEASE="25.05"
LSB_VERSION="25.05 (Warbler)"
Running under: X11
OpenGL: '4.6 (Core Profile) Mesa 25.0.7' Detected version: 4.6
Frozen: False
Fonts:
  medium: GeistMonoNFM-Regular: /nix/store/sw6428v75mf2hznky24xcg7w6bzn3wrm-nerd-fonts-geist-mono-3.4.0+1.401/share/fonts/opentype/NerdFonts/GeistMono/GeistMonoNerdFontMono-Regular.otf:0
          Features: ()
    bold: GeistMonoNFM-SemiBold: /nix/store/sw6428v75mf2hznky24xcg7w6bzn3wrm-nerd-fonts-geist-mono-3.4.0+1.401/share/fonts/opentype/NerdFonts/GeistMono/GeistMonoNerdFontMono-SemiBold.otf:0
          Features: ()
  italic: GeistMonoNFM-Medium: /nix/store/sw6428v75mf2hznky24xcg7w6bzn3wrm-nerd-fonts-geist-mono-3.4.0+1.401/share/fonts/opentype/NerdFonts/GeistMono/GeistMonoNerdFontMono-Medium.otf:0
          Features: ()
      bi: GeistMonoNFM-SemiBold: /nix/store/sw6428v75mf2hznky24xcg7w6bzn3wrm-nerd-fonts-geist-mono-3.4.0+1.401/share/fonts/opentype/NerdFonts/GeistMono/GeistMonoNerdFontMono-SemiBold.otf:0
          Features: ()
Paths:
  kitty: /nix/store/dyyifi0w2g0ddfxv5xs6ihkgbr435qp2-kitty-0.42.1/bin/kitty
  base dir: /nix/store/dyyifi0w2g0ddfxv5xs6ihkgbr435qp2-kitty-0.42.1/lib/kitty
  extensions dir: /nix/store/dyyifi0w2g0ddfxv5xs6ihkgbr435qp2-kitty-0.42.1/lib/kitty/kitty
  system shell: /run/current-system/sw/bin/bash
System color scheme: dark. Applied color theme type: none
Loaded config files:
  /home/simon/.config/kitty/kitty.conf

Config options different from defaults:
font_family          family="GeistMono Nerd Font Mono"
Added shortcuts:
	ctrl+f →  push_keyboard_mode copy_text
	ctrl+i →  push_keyboard_mode insert_text
	ctrl+o > u →  homerow --type url --program default
	ctrl+shift+space →  kitten kitty-grab/grab.py
	ctrl+space →  send_text normal git status\r
Changed shortcuts:
	kitty_mod+u →  no_op # To avoid habits from alacritty
Changes in keyboard mode: copy_text
Added shortcuts:
	d →  homerow --type regex --regex "\\d{4}-\\d{2}-\\d{2}(?:[ T]\\d{2}:\\d{2}:\\d{2}(?:\\.[\w]+))?" --program @
	escape →  pop_keyboard_mode
	f →  homerow --type path --program @
	h →  homerow --type hash --program @
	i →  homerow --type regex --regex "[0-9a-f]{8}-[0-9a-f]{4}-[0-9a-f]{4}-[0-9a-f]{4}-[0-9a-f]{12}" --program @
	l →  homerow --type line --program @
	u →  homerow --type url --program @
	w →  homerow --type word --program @
Changes in keyboard mode: insert_text
Added shortcuts:
	d →  homerow --type regex --regex "\\d{4}-\\d{2}-\\d{2}(?:[ T]\\d{2}:\\d{2}:\\d{2}(?:\\.[\w]+))?" --program -
	escape →  pop_keyboard_mode
	f →  homerow --type path --program -
	h →  homerow --type hash --program -
	i →  homerow --type regex --regex "[0-9a-f]{8}-[0-9a-f]{4}-[0-9a-f]{4}-[0-9a-f]{4}-[0-9a-f]{12}" --program -
	l →  homerow --type line --program -
	shift+f →  homerow --multiple --type path --program -
	shift+i →  homerow --multiple --type regex --regex "[0-9a-f]{8}-[0-9a-f]{4}-[0-9a-f]{4}-[0-9a-f]{4}-[0-9a-f]{12}" --program -
	shift+l →  homerow --multiple --type line --program -
	u →  homerow --type url --program -
	w →  homerow --type word --program -
Colors:
	background           #202020   
	color0               #151515   
	color1               #ac4142   
	color10              #7e8d50   
	color11              #e5b566   
	color12              #6c99ba   
	color13              #9e4e85   
	color14              #7dd5cf   
	color15              #f5f5f5   
	color2               #7e8d50   
	color3               #e5b566   
	color4               #6c99ba   
	color5               #9e4e85   
	color6               #7dd5cf   
	color7               #d0d0d0   
	color8               #505050   
	color9               #ac4142   
	cursor               #d0d0d0   
	foreground           #d0d0d0   
	selection_background #303030   
	selection_foreground #202020   

Important environment variables seen by the kitty process:
	LANG                                en_US.UTF-8
	EDITOR                              vim
	SHELL                               /run/current-system/sw/bin/bash
	DISPLAY                             :0
	USER                                simon
	XDG_MENU_PREFIX                     plasma-
	LC_MONETARY                         de_DE.UTF-8
	XDG_SEAT                            seat0
	XDG_SESSION_DESKTOP                 KDE
	XDG_SESSION_TYPE                    x11
	XDG_CURRENT_DESKTOP                 KDE
	XDG_SEAT_PATH                       /org/freedesktop/DisplayManager/Seat0
	XDG_SESSION_CLASS                   user
	XDG_VTNR                            2
	XDG_SESSION_ID                      2
	XDG_RUNTIME_DIR                     /run/user/1000
	LC_TIME                             de_DE.UTF-8
	LC_ALL                              en_US.UTF-8
</pre>
</p>
</details>

**Comment by buzden** [comment]
I experience behaviour this on 0.43.1 too

**Comment by buzden** [comment]
The problem and solution is described here: https://stackoverflow.com/questions/342093/ps1-line-wrapping-with-colours-problem

TLDR: when color code is not between `\[` and `\]`, then count of symbols is calculated incorrectly. Just put all your `\e[<whatever>m` and other escape codes in between `\[` and `\]`.

Sad thing is that programs like `rlwrap` does not do this

**Comment by ssimono** [comment]
@buzden thanks for figuring this out 🙏 I confirm it solves the problem

**Comment by WuvQOS** [comment]
/wsh -n blow me if you insist. . . .
